### DISCUSSION TITLE
Is there a way to import util function into netlify edge function?

### DISCUSSION THREAD

**Discussion by dbzx10299** [nuxt/nuxt#25438] [Questions] [answered]
Is there a way to import a utility function from the `utils` directory into a netlify edge function?

If there‘s a function `sum` in `/utils/sum.js`

```js
export const sum = (…nums) => nums.reduce((prev, current) => prev + current)
```

And an edge function in `netlify/edge-functions/foo.js`

```js
const result = sum(1, 2, 3)

export default () => {
  return Response.json({ sum: result })
}

export const config = {
  path: '/foo'
}
```

This breaks and `sum` is not defined. I am able to import bare modules into the edge function but I can‘t import my own utility functions. 

I am also running `netlify dev` from the terminal since starting nuxt with `npm run dev` will not load the functions.

**Comment by TheAlexLichter** [comment]
You could define it as API endpoint in `server/api/sum.ts`, then it will be a fn when deploying to netlify!

**Reply by dbzx10299** [reply]
I use `npm run generate` and when I do this my functions in `/netlify/edge-functions` work but my server routes do not. The server routes do work though with `npm run build`

**Reply by TheAlexLichter** [reply]
Yes, that's correct. You could use npm run build and use e.g. the `prerender` route rule to prerender all pages but still use nitro!

**Reply by dbzx10299** [reply]
I submitted an issue for this before, I know it still hasn‘t been fixed but I cannot get the netlify contact form to work no matter what I do, so that‘s why I was using npm run generate for the blog. I did try what you said in the past and I remember the netlify crawlers did not pick up on the form, I even added it to the public directory and nothing I did worked.

**Reply by dbzx10299** [reply]
I don‘t want there to be an endpoint of `sum` though since its just a util, let‘s say we had a util function we used in many edge functions, if the util function was an endpoint we would have to make another network request, is there really no way to just use a util or any utility function we write in an edge function?

**Reply by TheAlexLichter** [reply]
@dbzx10299 actually no, request to other endpoints *from nitro endpoints* (so /api/result requests /api/sum) are no extra calls when using $fetch ☺️

**Reply by dbzx10299** [reply]
I have though in my example a netlify edge endpoint `/foo` and that returns a json object. The problem is that `sum()` defined somewhere else in my project cannot be used. I think I misunderstood what you meant. There is only one api endpoint here. 

Regardless whether `npm run build` or `npm run generate` is used my `netlify edge functions` work. I just don‘t know why I can‘t have a function `function doSomethingUseful() {}` defined somewhere else in the project and use it in the `edge function`

**Reply by TheAlexLichter** [reply]
Ahh sorry. Now I get it! Maybe https://github.com/nuxt/nuxt/issues/24841 helps you ;)

**Comment by dbzx10299** [comment, marked as answer]
I did get this to work finally, there are some important things to note here. 

- Direct Import
```js
import { sum } from '../../utils/sum.js'

export default async (req, context) => {
  const result = sum(2, 3, 4)
  return Response.json({ sum: result })
}
```
- Dynamic import
```js
export default async (req, context) => {
  // .js file extension must be included
  const { sum } = await import('../../utils/sum.js')
  const result = sum(2, 3, 4)
  return Response.json({ sum: result })
}
```

- Import maps

At the project root in `denos.json`
```js
{
  "imports": {
    "foo": "./utils/sum.js"
  }
}
```

Then at the root in `netlify.toml` we need the path to the import map
```
[functions]
  deno_import_map = "./deno.json"
```

Lastly in `netlify/edge-functions/foo.js`

```js
import { sum } from 'foo'
export default async (req, context) => {
  const result = sum(2, 3, 4)
  return Response.json({ sum: result })
}
```
